File: gcpc2016/oneway/submissions/accepted/egor_ek_vector.cpp
#include <unordered_map>
#include <vector>
#include <deque>
#include <algorithm>
#include <iostream>
#include <utility>

/*
  The same as egor_ek_better_start.cpp but uses std::vector and mapping to reverse edge rather than std::map
*/

struct CapEdge{
    size_t target;
    size_t reverse_edge;
    int cap;
    CapEdge(size_t target, size_t reverse_edge, int cap):
        target(target), reverse_edge(reverse_edge), cap(cap){}
};

typedef std::vector<CapEdge> Neighbors;//first->goal of the edge, second->capacitance
typedef std::vector<Neighbors> Graph;

typedef std::pair<size_t, int> Pred;//first->index of the reverse_edge, second->bottleneck value, min capacity
typedef std::vector<Pred> Preds;


size_t MAX=2000000000;
//remember the way and the minimal capacitance on the way 

bool bfs(const Graph &graph, size_t start, size_t end, Preds &preds){

    std::fill(preds.begin(), preds.end(), std::make_pair(MAX, MAX));
    std::deque<size_t> q;
    
    preds[start]=std::make_pair(2*MAX, MAX); 
    q.push_back(start);
    while (!q.empty()){
        size_t &cur=q.front();
        for(const auto &edge : graph[cur]){ 
            if(edge.cap>0 && preds[edge.target].first==MAX){
                preds[edge.target]=std::make_pair(edge.reverse_edge, std::min(preds[cur].second, edge.cap));
                q.push_back(edge.target);
                if(edge.target==end)
                    return true;
            }
        }
        q.pop_front();
   }
   return false;
}

int add_flow(Graph &graph, size_t end, const Preds &preds){
    int diff=preds[end].second;
    while(true){
        if(preds[end].second==MAX)
            break;
        CapEdge &back_edge=graph[end][preds[end].first];
        size_t next=back_edge.target;
        CapEdge &edge=graph[next][back_edge.reverse_edge];
        edge.cap-=diff;
        back_edge.cap+=diff;
        end=next;
    }
    return diff;
 }
    
int edmonds_karp(Graph &graph, size_t start, size_t end){
    int flow=0;
    size_t n=graph.size();
    Preds preds(n);
   
    while(bfs(graph, start, end, preds)){     
        flow+=add_flow(graph, end, preds);
    }

    return flow;
}
 
void add_edge(Graph &graph, size_t from, size_t to, int cap){
    CapEdge edge(to, graph[to].size(), cap);
    CapEdge rev_edge(from, graph[from].size(), 0);
    graph[from].push_back(edge);
    graph[to].push_back(rev_edge);
}


typedef std::pair<size_t, size_t> Edge;
int init_graph(Graph &graph, const std::vector<Edge> &edges, int d){  

    size_t n=graph.size();
    for(auto &aList:graph)
        aList.clear();
    
    std::vector<int> cnts(n,0);
    
    //try to minimize the number of outgoing edges in greedy manier:
    for(const auto &edge:edges){
         if(cnts[edge.first]<cnts[edge.second]){ 
             cnts[edge.first]++;
             add_edge(graph, edge.first, edge.second, 1);
         }
         else{
             cnts[edge.second]++;
             add_edge(graph, edge.second, edge.first, 1);
         }
    }
    
    int expected=0;
    for(size_t i=0;i<n-2;i++){
        if(cnts[i]>d){
            int needed=cnts[i]-d;
            add_edge(graph, n-2, i, needed);
            expected+=needed;
        }
        if(cnts[i]<d)
            add_edge(graph, i, n-1, d-cnts[i]);
    }
    
    return expected;
}

int estimate_max_d(const std::vector<Edge> &edges, int n){ 
    std::vector<int> cnts(n,0);
    for(const auto &edge:edges)
         if(cnts[edge.first]<cnts[edge.second])
             cnts[edge.first]++;
         else
             cnts[edge.second]++;
    
    return *std::max_element(cnts.begin(), cnts.end());
}

int main(){
    size_t n;
    size_t m;
    
    std::cin>>n>>m;
    std::vector<Edge> edges(m);
    
    for(size_t i=0;i<m;i++){
        std::cin>>edges[i].first>>edges[i].second;
        edges[i].first--;edges[i].second--;
   }

    Graph graph(n+2);
    int lower=0;
    int upper=estimate_max_d(edges, n);
    
    while(lower+1<upper){
        int middle=(lower+upper)/2;
        int expected=init_graph(graph, edges, middle);
        int received=edmonds_karp(graph, n, n+1);
        if(expected==received)
            upper=middle;
        else
            lower=middle;
    }
    std::cout<<upper<<"\n";
}

File: gcpc2016/oneway/submissions/accepted/egor_ek_vector.py
# translation of egor_ek_vector.cpp to python, max flow with edmonds karp

from collections import deque as Queue
from collections import namedtuple

#CapEdge = namedtuple('CapEdge', 'target reverse_edge cap')
WayEdge = namedtuple('WayEdge', 'reverse_edge min_cap')

MAX=2000000000

def bfs(graph, start, end):
    n=len(graph)
    preds=[(MAX, MAX)]*n
    q=Queue()
    
    preds[start]=WayEdge(2*MAX, MAX)
    q.append(start)
    while q:
        cur=q.popleft()
        for edge in graph[cur]:
            if edge[2]>0 and preds[edge[0]][0]==MAX:
                preds[edge[0]]=WayEdge(edge[1], min(preds[cur][1], edge[2]))
                q.append(edge[0])
                if edge[0]==end:
                    return preds
    return None
   
   

def add_flow(graph, end, preds):
    diff=preds[end].min_cap;
    while True:
        if preds[end].min_cap==MAX:
            break
        back_edge=graph[end][preds[end].reverse_edge]
        edge=graph[back_edge[0]][back_edge[1]]
        edge[2]-=diff
        back_edge[2]+=diff
        end=back_edge[0]
    return diff
    
def edmonds_karp(graph, start, end):
    flow=0 
    while True:
        preds=bfs(graph, start, end)
        if preds is None:
            break     
        flow+=add_flow(graph, end, preds)
    return flow

 
def add_edge(graph, start, end, cap):
    edge=[end, len(graph[end]), cap]
    rev_edge=[start, len(graph[start]), 0]
    graph[start].append(edge)
    graph[end].append(rev_edge)


def init_graph(node_cnt, edges, d):  
    n=2*node_cnt+2
    graph=[list() for _ in range(n)]
    cnts=[0]*n
    for edge in edges:
        if cnts[edge[0]]<cnts[edge[1]]:
            cnts[edge[0]]+=1
            add_edge(graph, edge[0], edge[1], 1)
        else:
            cnts[edge[1]]+=1
            add_edge(graph, edge[1], edge[0], 1)
    
    expected=0
    for i in range(n-2):
        if cnts[i]>d:
            needed=cnts[i]-d
            add_edge(graph, n-2, i, needed)
            expected+=needed
        if cnts[i]<d:
            add_edge(graph, i, n-1, d-cnts[i])
 
    return (expected, graph, n-2, n-1)


def estimate_max_d(edges, n):
    cnts=[0]*n
    for edge in edges:
         if cnts[edge[0]]<cnts[edge[0]]:
             cnts[edge[0]]+=1
         else:
             cnts[edge[1]]+=1
    return max(cnts)

#main:
n,m=list(map(int, [input(), input()]))
edges=[]
for _ in range(m):
    edges.append([x-1 for x in map(int, input().split())])
  

lower=0
upper=estimate_max_d(edges, n)
    
while lower+1<upper:
    middle=(lower+upper)//2;
    expected, graph, source, target=init_graph(n, edges, middle)
    received=edmonds_karp(graph, source, target)
    if expected==received:
        upper=middle
    else:
        lower=middle

print(upper)


